### DISCUSSION TITLE
Numpy

### DISCUSSION THREAD

**Discussion by coppermouse** [pygame-web/pygbag#103] [Ideas]
I would like to thank you for this project! I works very well and now it is a lot easier for me to deploy my games!

That being said I am still in need of more features.

I noticed that pygame.PixeArray is not available nor does any other numpy related things be.

Would it be possible to get numpy to work? I can not import it and numpy is very useful when optimizing games.

**Comment by FinFetChannel** [comment]
Have you seen this?
https://github.com/pygame-web/pygbag/issues/86

basically, you have to `import pygame.surfarray` too, i think its similar or other numpy dependent modules

**Reply by coppermouse** [reply]
Thanks! I missed it because ~~I forgot to check closed tickets.~~ I forgot to check Issues.

Ok! Numpy seems to work which surprised me because I am sure I checked it before and other people hinted it was hard to get numpy-related things to work in pygbag. That is good! PixelArray is secondary on my list.

I would also assume I need surfarray some day so now I know how to make it available as well.

That being said, when I run this code:

```
print("import pygame...")
import pygame
print("...done")


print("import numpy...")
import numpy
print("...done")


print("import pygame.surfarray...")
import pygame.surfarray
print("...done")

try:
    print("import PixelArray")
    import pygame.PixelArray
    print("...done")
except Exception as e:
    print(str(e))

print("try make pixel Array")
pygame.PixelArray(pygame.Surface((1,1)))
print("...done")
```

I got this result (using debug running pygbag)

```
import pygame...
...done
import numpy...
181: FIXME dependency table for manually built modules
1202: want='numpy' found : True
pkg : http://localhost:8000/archives/repo/pkg/numpy-1.22.4-cp311-cp311-wasm32_mvp_emscripten.whl
 |XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX| 100.0% 
resume=<function <module> at 0xe31450> is not awaitable
import pygame...
...done
import numpy...
...done
import pygame.surfarray...
...done
import PixelArray
No module named 'pygame.PixelArray'
try make pixel Array
Traceback (most recent call last):
  File "/data/data/org.python/assets/site-packages/pygame/__init__.py", line 50, in _attribute_undefined
    raise RuntimeError(f"{name} is not available")
RuntimeError: pygame.PixelArray is not available
```

There is still issues with PixelArray...

I also tested moving import numpy to top but it has the same results

**Reply by pmp-p** [reply]
pygame.PixelArray() is a constructor, it should be available right after `import pygame.surfarray`

**Reply by coppermouse** [reply]
I am afraid it isn't, do you get the another result than me when you run the code above in a pygbag session? I think I am running 0.7.1 (downloaded from pip)

**Reply by pmp-p** [reply]
https://pyga.me/docs/ref/pixelarray.html#pygame.PixelArray

and from [REPL](https://pygame-web.github.io/showroom/python.html?-i)
```
>>> import numpy
1204: want='numpy' found : True
pkg : https://pygame-web.github.io/archives/repo/pkg/numpy-1.22.4-cp311-cp311-wasm32_mvp_emscripten.whl
 |XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX| 100.0% 
resume=<function <module> at 0x63e388> is not awaitable
>>> import pygame
1204: want='pygame.base' found : True
pkg : https://pygame-web.github.io/archives/repo/pkg/pygame_static-1.0-cp311-cp311-wasm32_mvp_emscripten.whl
 |XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX| 100.0% 
resume=<function <module> at 0x866550> is not awaitable
https://github.com/pygame-web/pygbag/issues/16
    applying: use aio green thread for pygame.time.set_timer










sys._emscripten_info(emscripten_version=(3, 1, 29), runtime='Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36', pthreads=False, shared_memory=False)
pygame 2.1.4.dev1 (SDL 2.24.2, Python 3.11.1)
Hello from the pygame community. https://www.pygame.org/contribute.html
>>> import pygame.surfarray
>>> pygame.PixelArray
<function PixelArray at 0x18c9b48>
```

**Reply by coppermouse** [reply]
but you didn't actually call the PixelArray. I did what you just did on REPL but in the end I also called PixelArray()
```
>>> pygame.PixelArray(pygame.Surface((1,1)))
Traceback (most recent call last):
  File "/data/data/org.python/assets/site-packages/pygame/__init__.py", line 50, in _attribute_undefined
    raise RuntimeError(f"{name} is not available")
RuntimeError: pygame.PixelArray is not available
```

This is not a big deal for me really, I can always find-and-replace colors on a surface using a for-loop. Just let you know that PixelArray is not available according to error message

**Reply by pmp-p** [reply]
Then it's a bug in wasm pygame port that need to be reported on pygame-ce issues with the above reproducer

**Reply by coppermouse** [reply]
Ok. I trust you on this one and post this as an issue on pygame-ce.

Thanks! 

I guess I close this conversation because we come to some conclusion

**Comment by coppermouse** [comment]
I got answer on how to get numpy and surfarray to work but not PixelArray.

I keep this post Open some more time before I close it.

**Reply by oddbookworm** [reply]
Have you tried importing `pygame.pixelarray`? The import is not `pygame.PixelArray`

**Reply by pmp-p** [reply]
indeed pygame.pixelarray is missing from pygame-ce wasm build up to tested 2.3.1-dev1
so https://github.com/pygame-community/pygame-ce/issues/2283 is a valid issue.
